Picosoc fusesoc

olofk opened the pull request:
This adds three FuseSoC .core files for the spiflash model, the generic picosoc code and the hx8kdemo implementation of picosoc. The intention is to make it easier to reuse the generic parts of picosoc and the spiflash code in other projects.

The most controversial change here is splitting out picosoc_regs into a separate file. This was done both to not having to compile picosoc.v before picorv32.v as well as making it easier to source other implementations of picosoc_regs without having to edit any code

cliffordwolf commented:
Why are the .core files in the top level directory and not the picosoc subdirectory?

> The most controversial change here is splitting out picosoc_regs into a separate file.

I would have preferred that in a separate PR then. (But it's okay this time. I don't find this very controversial actually. :)

olofk commented:
> Why are the .core files in the top level directory and not the picosoc subdirectory?

Once FuseSoC has found a core in a directory, it is designed to stop looking for cores in subdirectories. Since there already is a picorv32.core in the root dir (I wonder who put it there :)), it would not be able to find the ones in the picosoc dir. This hasn't generally been a problem, but the picorv32 repo layout is a bit different. I was planning to explain why it needs to be done this way, but I really can't find a good reason for it. Maybe I had a great reason when I added this behaviour June 26, 2014, but I can't think of it now.

So here's my idea: If you don't mind having them in the toplevel, please apply the patches. Otherwise, I'll change the way FuseSoC traverses directories and I'll make a new PR once I have made the changes, and split out the two other patches in separate PRs

cliffordwolf commented:
Would it be possible to move them all in a fusesoc/ subdirectory?

olofk commented:
> Would it be possible to move them all in a fusesoc/ subdirectory?

Referencing files in the parent directory breaks the sandboxing of cores. With that said, it might work, but it's not supposed to :)

It did however give me an idea to support setting directories relative to the .core file as the root directory. If you don't want the files in the repo root, I'll open a new PR once I have made some changes to FuseSoC

cliffordwolf commented:
> It did however give me an idea to support setting directories relative to the .core file as the root directory. If you don't want the files in the repo root, I'll open a new PR once I have made some changes to FuseSoC

Yes, please. I'd like to keep the number of files in the root directory to a reasonable minimum.

I'd be fine with either a `fusesoc/` directory with all the .core files, or a picorv32.core in the top level and all picosoc related core files in `picosoc/`.

olofk commented:
Sure thing. Closing for now
